### ISSUE TITLE
Great start, but not 100% protocol compatible 

### ISSUE THREAD

**Issue by netmindz** [chrisgott/feed_my_wled#1]
Great to see more folk getting involved with the Audio Responsive stuff and nice to have an option for non-windows 

You are missing some of the V2 data however in your current implementation 

See https://mm.kno.wled.ge/soundreactive/sync/#v2-format-wled-version-0140-including-moonmodules-fork

**Comment by chrisgott** [comment]
Hello. I guess you mean the empty Bytes. Some sources say they are used, some say they are for further usage. So i checked source  on actual version of WLED itself and found out they are definitly unused. Thats why. 

#this is from the source code of wled!
    struct __attribute__ ((packed)) audioSyncPacket {  // "packed" ensures that there are no additional gaps
      char    header[6];      //  06 Bytes  offset 0
      uint8_t reserved1[2];   //  02 Bytes, offset 6  - gap required by the compiler - not used yet 
      float   sampleRaw;      //  04 Bytes  offset 8  - either "sampleRaw" or "rawSampleAgc" depending on soundAgc setting
      float   sampleSmth;     //  04 Bytes  offset 12 - either "sampleAvg" or "sampleAgc" depending on soundAgc setting
      uint8_t samplePeak;     //  01 Bytes  offset 16 - 0 no peak; >=1 peak detected. In future, this will also provide peak Magnitude
      uint8_t reserved2;      //  01 Bytes  offset 17 - for future extensions - not used yet
      uint8_t fftResult[16];  //  16 Bytes  offset 18
      uint16_t reserved3;     //  02 Bytes, offset 34 - gap required by the compiler - not used yet
      float  FFT_Magnitude;   //  04 Bytes  offset 36
      float  FFT_MajorPeak;   //  04 Bytes  offset 40
};

Chris

**Comment by netmindz** [comment]
The version of the usermod in the "official" WLED is about 2 years old Vs the official source of that usermod which is the MoonModules repo.

**Comment by softhack007** [comment]
Hi @chrisgott, nice to see some external tools are growing around the WLED audio sync feature.

@netmindz is right, the official source for the protocol is the [MoonModules fork of WLED](https://github.com/MoonModules/WLED), plus stuff that's in my head and in the source code (i'm the main author of WLED audioreactive) 😀 


```C++
    // new "V2" audiosync struct - 44 Bytes
    struct __attribute__ ((packed)) audioSyncPacket {  // WLEDMM "packed" ensures that there are no additional gaps
      char    header[6];      //  06 Bytes  offset 0
      uint8_t pressure[2];    //  02 Bytes, offset 6  - sound pressure as fixed point (8bit integer,  8bit fraction) 
      float   sampleRaw;      //  04 Bytes  offset 8  - either "sampleRaw" or "rawSampleAgc" depending on soundAgc setting
      float   sampleSmth;     //  04 Bytes  offset 12 - either "sampleAvg" or "sampleAgc" depending on soundAgc setting
      uint8_t samplePeak;     //  01 Bytes  offset 16 - 0 no peak; >=1 peak detected. In future, this will also provide peak Magnitude
      uint8_t frameCounter;   //  01 Bytes  offset 17 - track duplicate/out of order packets
      uint8_t fftResult[16];  //  16 Bytes  offset 18
      uint16_t zeroCrossingCount; // 02 Bytes, offset 34
      float  FFT_Magnitude;   //  04 Bytes  offset 36
      float  FFT_MajorPeak;   //  04 Bytes  offset 40
    };
```

The ultimate source for information about how to fill the data structure is the source code, especially this part:

https://github.com/MoonModules/WLED/blob/f24d35e970ec04359d51d0d37ce9c0d12da381db/usermods/audioreactive/audio_reactive.h#L1696-L1732

> I'm complaining that the WARLS2 protocol isn't specified in enough detail (e.g., what is sound pressure, which unit is used, how to calculate it, what is peak magnitude, and similar aspects).

Actually I would not call it " WARLS2 protocol", as it was intended for communication between WLED instances. So the content of the structure is driven by WLED internal data structures.

The minimal content you need is:
* sample magnitude = biggest audio sample over 25 milliseconds, scaled so the max value is ~255
* FFT results, put into 16 "channels" that are based on an FFT analysis with 512 samples. Code is here: https://github.com/MoonModules/WLED/blob/f24d35e970ec04359d51d0d37ce9c0d12da381db/usermods/audioreactive/audio_reactive.h#L855-L872
* FFT magnitude = value of biggest FFT result from a single FFT run
* FFT MajorPeak = frequency (in Hz) of the strongest FFT result bin

Actually - for a start - you can set the complete data structure to zero, and only fill in `sampleSmth` and `sampleRaw`. This means some effects (especially the freqency reactive ones like GEQ) will show nothing, but its enough to validate the data transmission (use WLED "Gravimeter" effect to check how the esp32 device interprets your data).

Its also important that you send no more than 50 audioSyncPacket messages per second. More data will cause overflow of the internal receiver queue.

**Comment by softhack007** [comment]
PS: if you are wondering "why is the _official_ WLED about two years behind" ? 
Well complicated story. 

Its mainly due to licensing, as I had some very frustrating experiences with MIT license, so I was not willing to publish my advanced source code (MoonModules WLED fork) under MIT. [Aircoookie/WLED](https://github.com/Aircoookie/WLED) has moved away from MIT licensing with their latest release 0.15.0, so the additional features available today in [MoonModules WLED](https://github.com/MoonModules/WLED/tree/mdev) will come back to "upstream" WLED within 2025.

MoonModules WLED audioreactive supports the reduced `audioSyncPacket` format that comes from the older audioreactive code, but to be "future-safe" i'd suggest you mainly look at the code in MoonModules WLED.

**Comment by troyhacks** [comment]
> * `zeroCrossingCount` is the number of zero crossings observed in 25ms. Its only used by the [auto-playlist usermod](https://github.com/MoonModules/WLED/tree/mdev/usermods/usermod_v2_auto_playlist).

This is "only" used in the AutoPlaylist usermod, but it's integral to the functionality. 😃 

I would love to see it working with feed_my_wled

**Comment by softhack007** [comment]
> I would love to see it working with feed_my_wled

@chrisgott this value is very simple to calculate:
* if ( sign(lastSample) != sign(currentSample) ) zeroCrossingCount ++;
* reset each 512 samples

run this while pulling raw samples out of your audio input stream. Its best to actually apply a "DC blocker" (aka low-cut filter) to the audio data stream, to make sure that audio data is centered around 0.

 You can use this code for filtering the raw samples: 
* https://github.com/MoonModules/WLED/blob/f24d35e970ec04359d51d0d37ce9c0d12da381db/usermods/audioreactive/audio_reactive.h#L470-L484
* or this function that also prevents high frequency noise https://github.com/MoonModules/WLED/blob/f24d35e970ec04359d51d0d37ce9c0d12da381db/usermods/audioreactive/audio_reactive.h#L935-L964

**Comment by chrisgott** [comment]
Hi,

Thank you, @softhack007, for joining the discussion and shedding light on this matter. It's a great honor for me to meet the main author on my very first GitHub experience.

I now understand why I found multiple definitions of the protocol—I'll definitely take these features into account for the next version of my script. Interestingly, I had already implemented them during development, but they didn’t work. After discovering they weren't considered in WLED 0.15, I removed them again.

I’ve understood the required parameters, and I hope my technical understanding is sufficient to implement them. This is not only my first GitHub project but also my first real foray into Python (I come from PHP development in the ’90s…).

My stream is designed to handle everything—from low-fi to high-end sources. In its current setup, I have minimal control over levels and rates because I consider the pipe to be fully driven. However, I’ll add AGC for future-proofing. To minimize the number of packets, the script reads 8 kB chunks at a time (configurable), which brings me to around 25 packets per second—a target I was aiming for. In my opinion, 25 PPS is sufficient for a smooth experience.

I’m looking forward to version 3 of the protocol, as I see that v2 may not cover everything the future might demand. I’m imagining cool animations working with peaks, levels, drops, and more for each band, with (subdivided) beats for rhythmic light shows. That’s exactly why I wanted to bring my WLED in the basement to life with music. Party party! 🎉

Alright, time to sit down and do my homework. I’ll let you know once v0.2 is ready.

One (serious) question remains: Since audio_reactive.h already handles all the analysis for microphone input (including fft calculations), why doesn’t it simply accept an (even minimized) audio stream?

Best regards,
Christian

**Comment by softhack007** [comment]
> question remains: Since audio_reactive.h already handles all the analysis for microphone input (including fft calculations), why doesn’t it simply accept an (even minimized) audio stream?

That's a good question - actually the sound sync protocol takes the result of all audio processing done by the esp32, and transmits that to another esp32. If you imagine an audio pipeline from microphone towards effects, our protocol "hooks in" at the end of the pipeline.

To accept an audio stream directly, we would need to rep,have the microphone/line-in sampling code with a stub that receives from the network. This has not been tried yet, but I see some possible troubles with it:
* latency - our audio processing code needs a continuous data stream from the audio source. Any "gap" will be visible as random noise in effects
* volume of data - little esp32 has very limited receive buffers. At 22khz, we would expect to receive 2048 bytes each 12 milliseconds - this puts the device under stress and it will slow down led effects.
* control over sampling parameter - our audio processing pipeline can only handle 22khz sampling rate, 32bit samples provided as signed float.

So our decision was to no go this way, but only tansfer processed result to other devices - it also brings down the amount of data needed in 20ms (from 2048 bytes to only 44 bytes).

Btw, it looks like you send the UDP data to the esp32 ip address directly? I'm a bit surprised that this works. Actually sound sync uses a multicast group (managed by the router) at `239.0.0.1`. The idea was that all devices (send or receive) join this multicast group to access the audiosync data, instead of building direct one-by-one connections.


Happy new year / einen guten Rutsch ins neue Jahr 🥳 
Frank

**Comment by softhack007** [comment]
> For example, you mention value of biggest FFT result from a single FFT run, but FFT codes differ in how they present "values". Are "FFT values" the value in a specific channel in Volts (from a hypothetical mic), the amplitude in Volts^2, or perhaps a decibel value, relative to some fiducial amplitude?

Hi @jdtsmith, our code does not use any physical units like volts, decibel, watts etc.

The expected value ranges are already described in the official wiki https://mm.kno.wled.ge/soundreactive/sync/#values


Effects show best reactions if `sampleRaw`, `sampleSmth` and `fftResult[]` use the full ranges of [0.0 ... 255.0] - maybe -10% so you have some room for peaks. `sampleSmth = 0.0` is interpreted as "silence".

This is one reason why you might need your own AGC implementation, to adjust scaling factors based on your current audio input characteristics.

For the case of `fft_Magnitude`, I can only repeat what I said before - Typical range is [0... 4096], in units that only the ArduinoFFT library knows 🤷.

**Comment by softhack007** [comment]
> I'm also planning (again) a small tool to parse an AirPlay stream and forward via multicast to WLED groups. I'll probably go with a much simplified route at first to avoid all the complexities.

There is one thing worth mentioning, when you work with your raw audio data stream to extract data for WLED audio sync.

 Audio drivers often deliver samples as 16bit, 24bit or 32bit *unsigned* integer. But when looking for the "max sample in 25ms", you need to use *signed samples* where the centre value (silence) is zero, and amplitudes go both into positive and negative direction. Then track the max absolute sample amplitude in 25ms, apply your scaling factor and send it to WLED in an audioSyncData package.

One way to achieve this "centered around zero" is to run a low-cut filter with a cut-off frequency below 40hz.

**Comment by jdtsmith** [comment]
> > For example, you mention value of biggest FFT result from a single FFT run, but FFT codes differ in how they present "values". Are "FFT values" the value in a specific channel in Volts (from a hypothetical mic), the amplitude in Volts^2, or perhaps a decibel value, relative to some fiducial amplitude?
> 
> Hi @jdtsmith, our code does not use any physical units like volts, decibel, watts etc.

Whether you use units explicitly or not, they are always there :).  The main point is that a mic turns a physical induced voltage into an amplitude value (positive or negative as the wave induces motion in both directions).  The power in the delivered sound wave at a given frequency is proportional to the square of that amplitude (always positive), and a logarithmic representation of the power (relative to some _reference value_) is expressed in decibels.  It sounds like AudioSync wants to deal directly in voltage=amplitude space (positive or negative values).  In the end these are just interpreted for various effects, so unit flavor only "matters" via that interpretation.  Just hoping for consistency with the internal usage.  

> The expected value ranges are already described in the official wiki https://mm.kno.wled.ge/soundreactive/sync/#values
> 
> Effects show best reactions if `sampleRaw`, `sampleSmth` and `fftResult[]` use the full ranges of [0.0 ... 255.0] - maybe -10% so you have some room for peaks. `sampleSmth = 0.0` is interpreted as "silence".

Right, and you are using "absolute amplitude" instead of "squared amplitude=power" as your "sample strength".

> For the case of `fft_Magnitude`, I can only repeat what I said before - Typical range is [0... 4096], in units that only the ArduinoFFT library knows 🤷.

:) .  Thanks.

**Comment by jdtsmith** [comment]
BTW, on this issue:

> - network latency - we found that the wifi protocol sometimes delivers a bunch of UDP packets in bursts, with 100ms (up to 250ms) between bursts. 

I've opened MoonModules/WLED#214.  From my experience with a network jitter smoothing algorithm for the RealTime UDP protocol (which sends even more data than the 44 Byte `audioSync`), this should work very well.  The basic idea is to trade latency for uniformity of playback cadence.  So when a bunch of packets come in with a long pause, the underlying code "plays back" from its internal buffer, smoothing things out.  Since network players can usually add delay to their audio output for proper sync, this predictable latency is easy to correct.

**Comment by Victoare** [comment]
@softhack007 many thanks for laying this out. I can finally understand (and implement) better some of those magic numbers :)
